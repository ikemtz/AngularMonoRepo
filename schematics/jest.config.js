module.exports = {
  testMatch: ["**/+(*.)+(spec|test).+(ts)?(x)"],
  transform: {
    "^.+\\.(ts|js|html)$": "ts-jest",
  },
  testPathIgnorePatterns: ["/node_modules/", "/dist/"],
  moduleFileExtensions: ["ts", "js", "html"],
  collectCoverage: true,
  coverageReporters: ["html", "lcov", "cobertura"],
  reporters: [
    "default",
    [
      "jest-junit",
      {
        outputDirectory: ".",
        outputName: "./junit.schematics.xml",
      },
    ],
  ],
  moduleNameMapper: {},
};
