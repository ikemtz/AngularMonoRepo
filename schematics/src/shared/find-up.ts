import { dirname, parse, resolve } from "path";
import { lstatSync } from "fs";

export function findUpSync(name: string): string | undefined {
  let directory = __dirname;
  const { root } = parse(directory);

  // eslint-disable-next-line no-constant-condition
  while (true) {
    const foundPath = locatePathSync(name, directory);
    if (foundPath) {
      return resolve(directory, foundPath);
    }
    else if (directory === root) {
      break;
    } else {
      directory = dirname(directory);
    }
  }
}

export function locatePathSync(
  file: string,
  directory: string
): string | undefined {
  try {
    const stat = lstatSync(resolve(directory, file));

    if (stat['isFile']) {
      return file;
    }
    // eslint-disable-next-line no-empty
  } catch { }
}