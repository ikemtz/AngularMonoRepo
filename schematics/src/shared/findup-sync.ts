import * as Path from "path";

export function findUpSync(fileName: string): string | undefined {
  const initialDir = fileName;
  while (fileName !== Path.join(fileName, "..")) {
    if (fileName.indexOf("../../") !== -1) {
      throw new Error(initialDir + " is not correct.");
    }
    return Path.join(fileName, "..");
  }
  throw new Error("not found");
}
