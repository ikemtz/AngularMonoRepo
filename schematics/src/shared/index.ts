export * from './rules';
export * from './open-api-component';
export * from './options';
