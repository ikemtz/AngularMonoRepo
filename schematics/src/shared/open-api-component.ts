export interface OpenApiComponent {
  properties: Properties;
  required?: string[];
}
export interface Properties {
  [key: string]: PropertyInfo;
}
export interface PropertyInfo {
  name?: string;
  enum?: [string | number];
  type: 'string' | 'boolean' | 'array' | 'integer' | 'number' | 'date';
  format?: 'date-time' | 'uuid' | 'int32' | 'double';
  nullable?: boolean;
  readOnly?: boolean;
  required: boolean;
  htmlInputType: 'text' | 'number' | 'checkbox' | 'date';
  snakeCaseName: string;
  maxLength: number;
  testFactoryValue: string;
}
