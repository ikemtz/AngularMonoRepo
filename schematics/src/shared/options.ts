import { PropertyInfo } from './open-api-component';

export interface IOptions {
  /** Entity Name (Must match schema name in json) */
  name: string;
  /** Parent Entity Name (Must match schema name in json) */
  parentName?: string;
  /** NGRX Store Feature Name */
  storeName?: string;
  /** Angular application Prefix */
  appPrefix: string;
  /** Folder to place generated files */
  path: string;
  /** Optional: Url of remote OpenApi JSON document (If this is not provided, then you must specify openApiJsonFileName) */
  openApiJsonUrl?: string;
  /** Optional: File path of local OpenApi JSON document (If this is not provided, then you must specify openApiJsonUrl) */
  openApiJsonFileName?: string;
  /** Optional: This is the property that will used for initial sorting on list and auto focus on crud schematics */
  modelFolderLocation?: string;
  /** Optional: Location of model files relative to the module/component generated root */

  firstProperty?: PropertyInfo;
  /** Ignore: This is for internal schematic use only  */
  swaggerProperties: PropertyInfo[];
  /** Ignore: This is for internal schematic use only  */
  swaggerObjectProperties?: PropertyInfo[];
  /** Ignore: This is for internal schematic use only  */
  hasDates?: boolean;
  hasObjects?: boolean;
  hasNullableDates?: boolean;
}
