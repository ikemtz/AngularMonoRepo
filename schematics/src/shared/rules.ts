import {
  apply,
  applyTemplates,
  chain,
  DirEntry,
  mergeWith,
  move,
  Rule,
  SchematicContext,
  Tree,
  url,
} from '@angular-devkit/schematics';
import { strings, normalize } from '@angular-devkit/core';
import fetch from 'node-fetch';
import { PropertyInfo, OpenApiComponent } from './open-api-component';
import * as pluralize from 'pluralize';
import { Observable, from, of } from 'rxjs';
import { map, switchMap } from 'rxjs/operators';
import { IOptions } from './options';
import _ = require('lodash');
import { TslintFixTask } from '@angular-devkit/schematics/tasks';
import * as fs from 'fs';
import * as findUp from 'find-up';
import * as https from 'https';
import * as http from 'http';

export function getSwaggerDoc(options: IOptions): Rule {
  return (host: Tree, context: SchematicContext) => {
    let jsonDoc: Observable<any> | null = null;
    if (options.openApiJsonFileName) {
      const fileName = getFileNames(options.openApiJsonFileName);
      context.logger.info(`Getting Swagger Document @${fileName}`);
      if (!fileName) {
        throw new Error(`File not found: ${options.openApiJsonFileName}`);
      }
      const response = fs.readFileSync(fileName);
      const apiDoc = JSON.parse(response.toString());
      jsonDoc = of(apiDoc);
    } else if (options.openApiJsonUrl) {
      context.logger.info(`Getting Swagger Document @${options.openApiJsonUrl}`);
      const httpsAgent = options.openApiJsonUrl.toLowerCase().startsWith("https") ? new https.Agent({
        rejectUnauthorized: false,
      }) : new http.Agent();
      jsonDoc = from(fetch(options.openApiJsonUrl, {
        agent: httpsAgent
      })).pipe(
        switchMap(resp => from(resp.json())));
    }
    if (jsonDoc) {
      return jsonDoc.pipe(map(data => processOpenApiDoc(data, options, host)));
    }
    return of(host);

  };
}

function getFileNames(openApiJsonFileName: string) {
  if (fs.existsSync(`${__dirname}/${openApiJsonFileName}`)) {
    return `${__dirname}/${openApiJsonFileName}`;
  }
  return findUp.sync(openApiJsonFileName);
}

export function processOpenApiDoc(data: any, schema: IOptions, host: Tree) {
  const openApiComonent = data.components.schemas[strings.classify(schema.name)] as OpenApiComponent;
  if (!openApiComonent) {
    throw new Error(`OpenApi Component not found in swagger doc: ${schema.name}`);
  }
  const properties = openApiComonent.properties as {
    [key: string]: PropertyInfo;
  };
  const filteredProperties: PropertyInfo[] = [];
  const excludedFields = ['createdBy', 'createdOnUtc', 'tenantId', 'updatedBy', 'updatedOnUtc'];
  schema.hasDates = false;
  for (const propertyKey in properties) {
    if (excludedFields.indexOf(propertyKey) < 0 && properties[propertyKey].type !== 'array') {
      const property =
        mapPropertyAttributes(schema, properties[propertyKey], {
          ...properties[propertyKey],
          name: propertyKey,
          required: (openApiComonent.required ?? []).indexOf(propertyKey) > -1,
        });
      filteredProperties.push(property);
    }
  }
  schema.swaggerProperties = filteredProperties;
  return host;
}

function mapPropertyAttributes(options: IOptions, source: PropertyInfo, dest: any) {

  if (source.type === 'number' || source.type === 'integer') {
    dest.htmlInputType = 'number';
    dest.filterExpression = 'numeric';
    dest.testFactoryValue = '0';
  } else if (source.type === 'boolean') {
    dest.htmlInputType = 'checkbox';
    dest.filterExpression = 'boolean';
    dest.testFactoryValue = 'true';
  } else if (source.format === 'date-time') {
    dest.htmlInputType = 'date';
    dest.filterExpression = 'date';
    dest.testFactoryValue = 'new Date()';
    options.hasDates = true;
  } else {
    dest.htmlInputType = 'text';
    dest.filterExpression = 'text';
    dest.testFactoryValue = `'${dest.maxLength ?
      _.snakeCase(dest.name).toUpperCase().substring(0, dest.maxLength) : _.snakeCase(dest.name).toUpperCase()}'`;
  }
  dest.snakeCaseName = _.snakeCase(dest.name);
  dest.startCaseName = _.startCase(dest.name);
  return dest;
}

export function generateFiles(schema: IOptions, templateType: 'list' | 'crud' | 'module'): Rule {
  return (tree: Tree, context: SchematicContext) => {
    context.logger.info(`Running ${strings.capitalize(templateType)} Schematic`);

    const singularizedName = pluralize(schema.name, 1);
    const pluralizedName = pluralize(schema.name, 2);
    const singularizedStoreName = schema.storeName ? pluralize(schema.storeName, 1) : '';
    const pluralizedStoreName = schema.storeName ? pluralize(schema.storeName, 2) : '';

    const folderName = normalize(`${schema.path}/${strings.dasherize(pluralizedName)}-${templateType}`);
    const templateSource = apply(url('./files'), [
      applyTemplates({
        ...strings,
        ...schema,
        pluralizedName,
        singularizedName,
        pluralizedStoreName,
        singularizedStoreName,
      }),
      move(folderName),
    ]);
    return chain([mergeWith(templateSource)])(tree, context);
  };
}

export function runLint(schema: IOptions): Rule {
  return (tree: Tree, context: SchematicContext) => {
    const dir: DirEntry | null = tree.getDir(schema.path.substr(0, schema.path.lastIndexOf('/')));
    const files = tree.actions.reduce((acc: Set<string>, action) => {
      const path = action.path.substr(1); // Remove the starting '/'.
      if (path.endsWith('.ts') && dir && action.path.startsWith(dir.path)) {
        acc.add(path);
      }
      return acc;
    }, new Set<string>());
    context.addTask(
      new TslintFixTask({
        ignoreErrors: false,
        tsConfigPath: 'tsconfig.json',
        files: Array.from(files),
      }),
    );
    return tree;
  };
}
