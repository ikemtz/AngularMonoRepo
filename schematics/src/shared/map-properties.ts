import * as pluralize from 'pluralize';
import { PropertyInfo, OpenApiComponent, OpenApiDocument } from './open-api-component';
import { IOptions } from './options';
import { mapPropertyAttributes } from './rules';

const excludedFields = ['createdBy', 'createdOnUtc', 'tenantId', 'updatedBy', 'updatedOnUtc'];

export function mapProperties(properties: { [key: string]: PropertyInfo; }, options: IOptions, openApiComponent: OpenApiComponent, openApiDoc: OpenApiDocument) {
  let firstProperty: PropertyInfo | undefined;
  const filteredProperties: PropertyInfo[] = [];
  for (const propertyKey in properties) {
    const originalProperty = properties[propertyKey];
    if (excludedFields.indexOf(propertyKey) < 0 && originalProperty.type !== 'array') {
      const property = mapPropertyAttributes(options, originalProperty, {
        ...properties[propertyKey],
        name: propertyKey,
        required: (openApiComponent.required ?? []).indexOf(propertyKey) > -1,
      });

      mapReferencedProperties(property, propertyKey, openApiComponent, openApiDoc);
      if (!firstProperty && propertyKey !== 'id') {
        firstProperty = property;
      }
      filteredProperties.push(property);
    }
  }
  return { firstProperty, filteredProperties };
}
function mapReferencedProperties(property: PropertyInfo, propertyKey: string, openApiComponent: OpenApiComponent, openApiDoc: OpenApiDocument) {
  const prop = openApiComponent.properties[propertyKey];
  const propTypeNames = prop['$ref']?.split('/');
  if (propTypeNames) {
    const typeName: string = propTypeNames.pop() || '';
    property.propertyTypeName = typeName;
    property.pluralizedPropertyTypeName = pluralize.plural(typeName);
    property.singularizedPropertyTypeName = pluralize.singular(typeName);
    const refComponent = openApiDoc.components.schemas[typeName];
    property.enum = refComponent.enum;
    if (property.enum) {
      const value = property.enum[0].toString().split(' ').pop();
      property.testFactoryValue = `${property.pluralizedPropertyTypeName}.${value}`
    }
  }
}

