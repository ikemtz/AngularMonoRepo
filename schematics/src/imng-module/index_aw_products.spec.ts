import { Tree } from '@angular-devkit/schematics';
import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import * as path from 'path';
import { IOptions } from '../shared';
import * as pluralize from 'pluralize'; 
import { classify, dasherize } from '@angular-devkit/core/src/utils/strings';

const collectionPath = path.join(__dirname, `../collection.json`);

describe(`imng-module`, () => {
  let tree: UnitTestTree;
  let options: IOptions;
  beforeAll(async () => {
    const runner = new SchematicTestRunner(`schematics`, collectionPath);
    options = {
      name: `product`,
      openApiJsonUrl: `https://awod-ikemtz.azurewebsites.net/swagger/v1/swagger.json`,
      path: `./test`,
      swaggerProperties: [],
      appPrefix: 'aw',
    };
    tree = await runner.runSchematic(`imng-module`, options, Tree.empty());
  })
  test(`tree files should match`, () => {
    expect(tree.files).toMatchSnapshot();
  });

  test(`module should work`, () => {
    const file = tree.get(
      `/test/products-module/products.module.ts`
    );
    const content = file?.content.toString();
    expect(content).toMatchSnapshot()
  });
  test(`list component should work`, () => {
    const file = tree.get(`/test/products-module/products-list/list.component.ts`);
    const content = file?.content.toString();
    expect(content).toMatchSnapshot();
  });
  test(`base crud component should work`, () => {
    const file = tree.get(`/test/products-module/products-crud/base-entry.component.ts`);
    const content = file?.content.toString();
    expect(content).toMatchSnapshot();
  });
  test(`crud facade template should work`, () => {
    const file = tree.get(`/test/products-module/products-crud/crud.facade.ts`);
    const content = file?.content.toString();
    expect(content).toMatchSnapshot();
  });
  test(`crud facade spec template should work`, () => {
    const file = tree.get(`/test/products-module/products-crud/crud.facade.spec.ts`);
    const content = file?.content.toString();
    expect(content).toMatchSnapshot();
  });
  test(`selectors template should work`, () => {
    const file = tree.get(`/test/products-module/+state/product.selectors.ts`);
    const content = file?.content.toString();
    expect(content).toMatchSnapshot();
  });
  test(`crud html template should work`, () => {
    const file = tree.get(`/test/products-module/products-crud/add-edit.component.html`);
    const content = file?.content.toString();
    expect(content).toMatchSnapshot();
  });
  test(`add component template should work`, () => {
    const file = tree.get(`/test/products-module/products-crud/add.component.ts`);
    const content = file?.content.toString();
    expect(content).toMatchSnapshot();
  });
  test(`edit component template should work`, () => {
    const file = tree.get(`/test/products-module/products-crud/edit.component.ts`);
    const content = file?.content.toString();
    expect(content).toMatchSnapshot();
  });
  test(`add component spec template should work`, () => {
    const file = tree.get(`/test/products-module/products-crud/add.component.spec.ts`);
    const content = file?.content.toString();
    expect(content).toMatchSnapshot();
  });
  test(`edit component spec template should work`, () => {
    const file = tree.get(`/test/products-module/products-crud/edit.component.spec.ts`);
    const content = file?.content.toString();
    expect(content).toMatchSnapshot();
  });
  test(`actions should work`, () => {
    const file = tree.get(`/test/products-module/+state/product.actions.ts`);
    const content = file?.content.toString();
    expect(content).toMatchSnapshot();
  });
  test(`reducers should work`, () => {
    const file = tree.get(`/test/products-module/+state/product.reducer.ts`);
    const content = file?.content.toString();
    expect(content).toMatchSnapshot();
  });
  test(`effects should work`, () => {
    const effectsFile = tree.get(`/test/products-module/+state/product.effects.ts`
    );
    const content = effectsFile?.content.toString();
    expect(content).toMatchSnapshot();
  });
  test(`list facade should work`, () => {
    const listFacadeSpecFile = tree.get(
      `/test/${pluralize(dasherize(options.name))}-module/${dasherize(
        pluralize(options.name)
      )}-list/list.facade.spec.ts`
    );
    const content = listFacadeSpecFile?.content.toString();
    expect(content).toContain(
      `expect(httpClient.get).toBeCalledWith('${dasherize(
        pluralize(options.name)
      )}-odata/odata/v1/${classify(pluralize(options.name))}?&$count=true');`
    );
  });
  test(`html file should work`, () => {
    const htmlFile = tree.get(`/test/products-module/products-list/list.component.html`
    );
    const content = htmlFile?.content.toString();
    expect(content).toMatchSnapshot();
  });
});
