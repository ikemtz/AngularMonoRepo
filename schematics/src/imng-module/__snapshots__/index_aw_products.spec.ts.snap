// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`imng-module actions should work 1`] = `
"import { createAction } from '@ngrx/store';
import { ODataResult, ODataState } from 'imng-kendo-odata';
import { createPayloadAction } from 'imng-ngrx-utils';

export const loadProductsRequest = createPayloadAction<ODataState>(
    '[Products] Load Products Request');
export const loadProductsSuccess = createPayloadAction<ODataResult<IProduct>>(
    '[Products] Load Products Success');
export const reloadProductsRequest = createAction(
    '[Products] Reload Products Request');
export const reloadProductsSuccess = createPayloadAction<ODataResult<IProduct>>(
    '[Products] Reload Products Success');

export const clearCurrentProduct = createAction('[Products] Clear Current Product');
export const setCurrentProduct = createPayloadAction<IProduct>('[Products] Set Current Product');
export const saveProductRequest = createPayloadAction<IProduct>('[Products] Save Product Request');
export const updateProductRequest = createPayloadAction<IProduct>('[Products] Update Product Request');
export const deleteProductRequest = createPayloadAction<IProduct>('[Products] Delete Product Request');


export const loadProductModelsRequest = createPayloadAction<ODataState>(
    '[Products] Load ProductModels Request');
export const loadProductModelsSuccess = createPayloadAction<ODataResult<IProductModel>>(
    '[Products] Load ProductModels Success');
export const loadProductCategoriesRequest = createPayloadAction<ODataState>(
    '[Products] Load ProductCategories Request');
export const loadProductCategoriesSuccess = createPayloadAction<ODataResult<IProductCategory>>(
    '[Products] Load ProductCategories Success');
"
`;

exports[`imng-module base crud component should work 1`] = `
"import { OnInit, Component } from '@angular/core';
import { BaseDataEntryComponent } from 'imng-kendo-data-entry';
import { BehaviorSubject, map, Observable, switchMap } from 'rxjs';

import { ProductCrudFacade } from './crud.facade';

@Component({ template: '' })
export abstract class ProductBaseEntryComponent extends BaseDataEntryComponent<ProductCrudFacade>
  implements OnInit {
  public readonly props = ProductProperties;  
  public readonly productModelProps = ProductModelProperties;
  public readonly productModels$: Observable<IProductModel[]>;
  public readonly productModelFilter$ = new BehaviorSubject('');  
  public readonly productCategoryProps = ProductCategoryProperties;
  public readonly productCategories$: Observable<IProductCategory[]>;
  public readonly productCategoryFilter$ = new BehaviorSubject('');

  constructor(facade: ProductCrudFacade) {
    super(facade);
    this.productModels$ = facade.productModels$.pipe(
      switchMap(productModels => this.productModelFilter$.pipe(
        map(productModelFilter => productModelFilter ? productModels
          .filter(productModel => (
            (productModel.name && productModel.name.toLowerCase().indexOf(productModelFilter) >= 0) ||
            (productModel.description && productModel.description.toLowerCase().indexOf(productModelFilter) >= 0)
          )) : productModels
    ))));
    this.productCategories$ = facade.productCategories$.pipe(
      switchMap(productCategories => this.productCategoryFilter$.pipe(
        map(productCategoryFilter => productCategoryFilter ? productCategories
          .filter(productCategory => (
            (productCategory.name && productCategory.name.toLowerCase().indexOf(productCategoryFilter) >= 0)
          )) : productCategories
    ))));
  }

  public ngOnInit(): void {
    this.initForm();
    this.facade.loadProductModels({ selectors: [
      ProductModelProperties.ID,
      ProductModelProperties.NAME,
      ProductModelProperties.DESCRIPTION,] });
    this.facade.loadProductCategories({ selectors: [
      ProductCategoryProperties.ID,
      ProductCategoryProperties.NAME,] });
  }

  public initForm(): void {
    this.addEditForm = ProductFormGroupFac();
  }

  public cancel(): void {
    this.facade.clearCurrentEntity();
  }

  public handleProductModelFilter(value: string) {
    this.productModelFilter$.next(value.toLowerCase());
  }
  public handleProductCategoryFilter(value: string) {
    this.productCategoryFilter$.next(value.toLowerCase());
  }
}
"
`;

exports[`imng-module crud facade template should work 1`] = `
"import { Injectable } from '@angular/core';
import { Store } from '@ngrx/store';
import { IDataEntryFacade } from 'imng-kendo-data-entry';
import { ODataState } from 'imng-kendo-odata';
import { productsFeature } from '../+state/product.reducer';
import { productQueries } from '../+state/product.selectors';
import * as productActionTypes from '../+state/product.actions';

@Injectable()
export class ProductCrudFacade implements IDataEntryFacade<IProduct> {
  loading$ = this.store.select(productsFeature.selectLoading);
  currentEntity$ = this.store.select(productQueries.selectCurrentProduct);
  isEditActive$ = this.store.select(productQueries.selectIsEditProductActive);
  isNewActive$ = this.store.select(productQueries.selectIsNewProductActive);
  productModels$ = this.store.select(productsFeature.selectProductModels);
  productCategories$ = this.store.select(productsFeature.selectProductCategories);

  constructor(private readonly store: Store) {}

  public setCurrentEntity(item: IProduct): void {
    this.store.dispatch(productActionTypes.setCurrentProduct(item));
  }

  public clearCurrentEntity(): void {
    this.store.dispatch(productActionTypes.clearCurrentProduct());
  }

  public saveNewEntity(item: IProduct): void {
    this.store.dispatch(productActionTypes.saveProductRequest(item));
  }

  public updateExistingEntity(item: IProduct): void {
    this.store.dispatch(productActionTypes.updateProductRequest(item));
  }

  public loadProductModels(state: ODataState): void {
    this.store.dispatch(productActionTypes.loadProductModelsRequest(state));
  }
  public loadProductCategories(state: ODataState): void {
    this.store.dispatch(productActionTypes.loadProductCategoriesRequest(state));
  }
}
"
`;

exports[`imng-module effects should work 1`] = `
"import { Injectable } from '@angular/core';
import { Store } from '@ngrx/store';
import { createEffect, Actions, ofType, concatLatestFrom } from '@ngrx/effects';
import { ODataService } from 'imng-kendo-odata';
import { handleEffectError } from 'imng-ngrx-utils';
import { map, switchMap } from 'rxjs/operators';

import { productsFeature } from './product.reducer';
import * as productActionTypes from './product.actions';
import { environment } from '../../../../environments/environment';

import { ProductApiService } from '../products-crud';

@Injectable()
export class ProductEffects {
  constructor(
    private readonly actions$: Actions,
    private readonly odataservice: ODataService,
    private readonly store: Store,
    private readonly productApiService : ProductApiService,
  ) {}

  loadProductsEffect$ = createEffect(() => {
    return this.actions$.pipe(
      ofType(productActionTypes.loadProductsRequest),
      switchMap((action: ReturnType<typeof productActionTypes.loadProductsRequest>) => this.odataservice
        .fetch<IProduct>(environment.odataEnpoints.products, action.payload)
        .pipe(
          map(t => productActionTypes.loadProductsSuccess(t)),
          handleEffectError(action))));
  });

  reloadProductsEffect$ = createEffect(() => {
    return this.actions$.pipe(
      ofType(productActionTypes.reloadProductsRequest),
      concatLatestFrom(() => this.store.select(productsFeature.selectGridODataState)),
      switchMap(([action, odataState]) => this.odataservice
        .fetch<IProduct>(environment.odataEnpoints.products, odataState, { bustCache: true })
        .pipe(
          map(t => productActionTypes.reloadProductsSuccess(t)),
          handleEffectError(action))));
  });

  saveProductEffect$ = createEffect(() => {
    return this.actions$.pipe(
      ofType(productActionTypes.saveProductRequest),
      switchMap((action: ReturnType<typeof productActionTypes.saveProductRequest>) => this.productApiService.post(action.payload).pipe(
        map(() => productActionTypes.reloadProductsRequest()),
        handleEffectError(action))));
  });

  updateProductEffect$ = createEffect(() => {
    return this.actions$.pipe(
      ofType(productActionTypes.updateProductRequest),
      switchMap((action: ReturnType<typeof productActionTypes.updateProductRequest>) => this.productApiService.put(action.payload).pipe(
        map(() => productActionTypes.reloadProductsRequest()),
        handleEffectError(action))));
  });

  deleteProductEffect$ = createEffect(() => {
    return this.actions$.pipe(
      ofType(productActionTypes.deleteProductRequest),
      switchMap((action: ReturnType<typeof productActionTypes.deleteProductRequest>) => this.productApiService.delete(action.payload).pipe(
        map(() => productActionTypes.reloadProductsRequest()),
        handleEffectError(action))));
  });

  loadProductModelsEffect$ = createEffect(() => {
    return this.actions$.pipe(
      ofType(productActionTypes.loadProductModelsRequest),
      switchMap((action: ReturnType<typeof productActionTypes.loadProductModelsRequest>) => this.odataservice
        .fetch<IProductModel>(environment.odataEnpoints.productModels, action.payload)
        .pipe(map(t => productActionTypes.loadProductModelsSuccess(t)),
          handleEffectError(action))));
  });

  loadProductCategoriesEffect$ = createEffect(() => {
    return this.actions$.pipe(
      ofType(productActionTypes.loadProductCategoriesRequest),
      switchMap((action: ReturnType<typeof productActionTypes.loadProductCategoriesRequest>) => this.odataservice
        .fetch<IProductCategory>(environment.odataEnpoints.productCategories, action.payload)
        .pipe(map(t => productActionTypes.loadProductCategoriesSuccess(t)),
          handleEffectError(action))));
  });
}
"
`;

exports[`imng-module list component should work 1`] = `
"import { Component, ChangeDetectionStrategy } from '@angular/core';
import { Router } from '@angular/router';
import { DetailExpandEvent } from '@progress/kendo-angular-grid';
import { KendoODataComponentBase } from 'imng-kendo-grid-odata';
import { ODataState } from 'imng-kendo-odata';

import { ProductListFacade } from './list.facade';
import { ProductCrudFacade } from '../products-crud';

const initialGridState: ODataState = {
  take: 20,
  skip: 0,
  selectors: [
    ProductProperties.ID,
    ProductProperties.NAME,
    ProductProperties.NUM,
    ProductProperties.COLOR,
    ProductProperties.STANDARD_COST,
    ProductProperties.LIST_PRICE,
    ProductProperties.SIZE,
    ProductProperties.WEIGHT,
    ProductProperties.PRODUCT_CATEGORY_ID,
    ProductProperties.PRODUCT_MODEL_ID,
    ProductProperties.SELL_START_DATE,
    ProductProperties.SELL_END_DATE,
    ProductProperties.DISCONTINUED_DATE,
    ProductProperties.THUMB_NAIL_PHOTO,
    ProductProperties.PRODUCT_MODEL,
    ProductProperties.PRODUCT_CATEGORY,
  ],
  sort: [
    { field: ProductProperties.NAME, dir: 'asc' },
  ],
  expanders: [
      { table: ProductProperties.PRODUCT_MODEL,
        selectors: [
          ProductModelProperties.ID,
          ProductModelProperties.NAME,
          ProductModelProperties.DESCRIPTION,
        ]},
      { table: ProductProperties.PRODUCT_CATEGORY,
        selectors: [
          ProductCategoryProperties.ID,
          ProductCategoryProperties.NAME,
        ]},
  ]
};

@Component({
  selector: 'aw-product-list',
  templateUrl: './list.component.html',
  styleUrls: ['./list.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ProductListComponent extends KendoODataComponentBase<IProduct, ProductListFacade> {
  public readonly props = ProductProperties;
  public currentItem: IProduct | undefined;

  constructor(facade: ProductListFacade, 
    public readonly crudFacade: ProductCrudFacade,
    router: Router) {
    super(facade, initialGridState, router);
  }

  public addItem(): void {
    this.crudFacade.setCurrentEntity({});
  }

  public editItem(item: IProduct): void {
    this.crudFacade.setCurrentEntity(item);
  }

  public deleteItem(item: IProduct): void {
    this.facade.deleteExistingEntity(item);
  }

  public detailExpanded(evt: DetailExpandEvent): void {
    this.currentItem = evt.dataItem;
  }
}
"
`;

exports[`imng-module module should work 1`] = `
"import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ReactiveFormsModule } from '@angular/forms';
import { StoreModule } from '@ngrx/store';
import { EffectsModule } from '@ngrx/effects';
import { GridModule, ExcelModule, PDFModule } from '@progress/kendo-angular-grid';
import { DialogModule } from '@progress/kendo-angular-dialog';
import { DateInputsModule } from '@progress/kendo-angular-dateinputs';
import { DropDownsModule } from '@progress/kendo-angular-dropdowns';
import { ImngKendoGridModule } from 'imng-kendo-grid';
import { ImngKendoGridODataModule } from 'imng-kendo-grid-odata';
import { ImngDataEntryDialogModule } from 'imng-kendo-data-entry';
import { ImngKendoGridFilteringModule } from 'imng-kendo-grid-filtering';

import { ProductsRoutingModule } from './products.routing';
import { productsFeature } from './+state/product.reducer';
import { ProductEffects } from './+state/product.effects';

import { ProductListComponent, ProductListFacade } from './products-list';
import { ProductAddComponent, ProductEditComponent, ProductApiService, ProductCrudFacade  } from './products-crud';


@NgModule({
  declarations: [ProductListComponent, ProductAddComponent, ProductEditComponent ],
  imports: [
    CommonModule,
    GridModule,
    ExcelModule,
    PDFModule,
    DialogModule,
    DateInputsModule,
    DropDownsModule,
    ImngDataEntryDialogModule,
    ImngKendoGridFilteringModule,
    ImngKendoGridModule,
    ImngKendoGridODataModule,
    ReactiveFormsModule,
    ProductsRoutingModule,
    StoreModule.forFeature(productsFeature),
    EffectsModule.forFeature([ProductEffects]),
  ],
  providers: [
    ProductListFacade,
    ProductCrudFacade,
    ProductApiService, 
  ],
})
export class ProductsModule { }
"
`;

exports[`imng-module reducers should work 1`] = `
"import { createReducer, on, createFeature } from '@ngrx/store';
import { createKendoODataGridInitialState, getODataPagerSettings, KendoODataGridState } from 'imng-kendo-grid-odata';
import { imngEffectError, imngEffectErrorReducer } from 'imng-ngrx-utils';

import * as productActionTypes from './product.actions';
export const PRODUCTS_FEATURE_KEY = 'products';

export interface State extends KendoODataGridState<IProduct> {
  currentProduct: IProduct | undefined;
  productModels: IProductModel[];
  productCategories: IProductCategory[];
}

export const initialState: State = {
  ...createKendoODataGridInitialState(),
  currentProduct: undefined,
  productModels: [],
  productCategories: [],
  loading: true,
};

export const productsFeature = createFeature({
  name: PRODUCTS_FEATURE_KEY,
  reducer: createReducer(
    initialState,
    on(productActionTypes.loadProductsRequest, 
      (state, { payload }) : State => ({
        ...state,
        gridODataState: payload,
        loading: true,
        error: null, })),
    on(productActionTypes.loadProductsSuccess,
      productActionTypes.reloadProductsSuccess, 
      (state, { payload }) : State => ({
        ...state,
        loading: false,
        gridPagerSettings: getODataPagerSettings({
          gridData: payload,
          gridODataState: state.gridODataState,
        }),
        gridData: payload,
        error: null, })),
    on(productActionTypes.setCurrentProduct, 
      (state, { payload }) : State => 
        ({ ...state, currentProduct: payload })),
    on(productActionTypes.clearCurrentProduct, 
      (state) : State => ({ ...state, currentProduct: undefined })),
    on(productActionTypes.saveProductRequest,
      productActionTypes.updateProductRequest,
      productActionTypes.deleteProductRequest,
      (state) : State => ({
        ...state,
        loading: true,
      }),
    on(productActionTypes.loadProductModelsSuccess,
      (state, { payload }): State => ({
        ...state,
        productModels: payload.data
      })),
    on(productActionTypes.loadProductCategoriesSuccess,
      (state, { payload }): State => ({
        ...state,
        productCategories: payload.data
      })),
    on(imngEffectError, imngEffectErrorReducer),
    ),
)
});
"
`;

exports[`imng-module tree files should match 1`] = `
Array [
  "/test/products-module/products.module.spec.ts",
  "/test/products-module/products.module.ts",
  "/test/products-module/products.routing.ts",
  "/test/products-module/+state/product.actions.ts",
  "/test/products-module/+state/product.effects.ts",
  "/test/products-module/+state/product.reducer.ts",
  "/test/products-module/+state/product.selectors.ts",
  "/test/products-module/products-list/index.ts",
  "/test/products-module/products-list/list.component.html",
  "/test/products-module/products-list/list.component.scss",
  "/test/products-module/products-list/list.component.spec.ts",
  "/test/products-module/products-list/list.component.ts",
  "/test/products-module/products-list/list.facade.spec.ts",
  "/test/products-module/products-list/list.facade.ts",
  "/test/products-module/products-crud/add-edit.component.html",
  "/test/products-module/products-crud/add-edit.component.scss",
  "/test/products-module/products-crud/add.component.spec.ts",
  "/test/products-module/products-crud/add.component.ts",
  "/test/products-module/products-crud/api.service.ts",
  "/test/products-module/products-crud/base-entry.component.ts",
  "/test/products-module/products-crud/crud.facade.spec.ts",
  "/test/products-module/products-crud/crud.facade.ts",
  "/test/products-module/products-crud/edit.component.spec.ts",
  "/test/products-module/products-crud/edit.component.ts",
  "/test/products-module/products-crud/index.ts",
]
`;
